### ISSUE TITLE
whether the SHAP will retrain the model which it evaluates? on the sub-dataset that contains part of features

### ISSUE THREAD

**Issue by dongdyang** [shap/shap#145]
I have a question about the SHAP.

After I train a model, eg. neural networks on my dataset, I use SHAP to evaluate.
Does SHAP use the same idea of the calculation of Shapley Value?

What I mean is the **SHAP will enumerate all combination of features** (human-understanding) on the training dataset but do not retrain the model waited for being evaluated on new sub-dataset (i am not sure whether it is on training or testing, it matters a lot!????), and get the shape value finally?

Is it the process of calculating the weight for each feature?

I will like to show a simple example to explain my question:
Assuming the dataset (age, gender, occupation), label is (employed, unemployed).
I use the Neural network as my model to do binary classification.
After i trained my model, i use SHAP to evaluate.

When I use the SHAP, all my features will be combined in different orders. It contains 8 orders: only [age], only [gender], only [occupation], [age, gender], [age, occupation], [gender, occupation], [age, gender, occupation]. 

In order to calculate the shapley value, the testing (or training? I'm confused here. A question raised here) inputs only containing sub-features like [gender, occupation] will be given the trained model. Then the prediction ratio will be the score of this feature combination [gender, occupation], right? This is my second question.

The SHAP will not train the model again based on sub-training (or sub-testing dataset)? This is my third question.

**Comment by slundberg** [comment]
I am not sure I fully understand the question. But the SHAP package does not retrain the model on the training set. Instead is uses sampling to approximate conditionals expectations with certain features held out. One (slow) way to do this is the add the features one-at-a-time for all possible orders (like you mentioned), and attribute the change in the model's expected output to each feature as it gets introduce (averaged over all possible orders). In practice we use faster methods and/or sampling approximations.

The training set that gets passed is used to estimate the conditional expectations. It does not need to be the test set. Also, if you have a deep model you might consider the DeepExplainer that we pushed this week.

**Comment by dongdyang** [comment]
You mentioned that "attribute the change in the model's expected output to each feature as it gets introduce (averaged over all possible orders)". 

Could I ask you one more question about how to get the feature importance in SHAP?

So, when you are getting the value of feature importance. You use the difference of the prediction result for two datasets [generated dataset and the original dataset] as the feature importance value, right???? Or you use the coefficient of the linear model to be the value of feature importance???


Thanks a lot!!!

**Comment by slundberg** [comment]
It turns out that taking a weighted average of the difference of predictions, and using the coefficient of a specially weighted linear model are the same. So both are true, just different ways of estimating the same thing. Kernel SHAP uses the linear regression way of computing it.
